rwengine/src/audio/SoundSource.cpp

```diff
@@ -0,0 +1,479 @@
+        avio_context_free(&formatContext->pb);
+        avcodec_close(codecContext);
+        avcodec_free_context(&codecContext);
+        avformat_close_input(&formatContext);
+        RW_ERROR("Couldn't open the audio codec context");
+        return;
+    }
+#endif
+
+    // Expose audio metadata
+    channels = static_cast<size_t>(codecContext->channels);
+    sampleRate = sdt.assetInfo.sampleRate;
+
+    // OpenAL only supports mono or stereo, so error on more than 2 channels
+    if (channels > 2) {
+        RW_ERROR("Audio has more than two channels");
+        av_free(formatContext->pb->buffer);
+        avio_context_free(&formatContext->pb);
+        av_frame_free(&frame);
+        avcodec_close(codecContext);
+        avformat_close_input(&formatContext);
+        return;
+    }
+
+    // Start reading audio packets
+    AVPacket readingPacket;
+    av_init_packet(&readingPacket);
+
+#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(57, 37, 100)
+
+    while (av_read_frame(formatContext, &readingPacket) == 0) {
+        if (readingPacket.stream_index == audioStream->index) {
+            AVPacket decodingPacket = readingPacket;
+
+            while (decodingPacket.size > 0) {
+                // Decode audio packet
+                int gotFrame = 0;
+                int len = avcodec_decode_audio4(codecContext, frame, &gotFrame,
+                                                &decodingPacket);
+
+                if (len >= 0 && gotFrame) {
+                    // Write samples to audio buffer
+
+                    for (size_t i = 0;
+                         i < static_cast<size_t>(frame->nb_samples); i++) {
+                        // Interleave left/right channels
+                        for (size_t channel = 0; channel < channels;
+                             channel++) {
+                            int16_t sample = reinterpret_cast<int16_t*>(
+                                frame->data[channel])[i];
+                            data.push_back(sample);
+                        }
+                    }
+
+                    decodingPacket.size -= len;
+                    decodingPacket.data += len;
+                } else {
+                    decodingPacket.size = 0;
+                    decodingPacket.data = nullptr;
+                }
+            }
+        }
+        av_free_packet(&readingPacket);
+    }
+#else
+
+    while (av_read_frame(formatContext, &readingPacket) == 0) {
+        if (readingPacket.stream_index == audioStream->index) {
+            AVPacket decodingPacket = readingPacket;
+
+            int sendPacket = avcodec_send_packet(codecContext, &decodingPacket);
+            int receiveFrame = 0;
+
+            while ((receiveFrame =
+                        avcodec_receive_frame(codecContext, frame)) == 0) {
+                // Decode audio packet
+
+                if (receiveFrame == 0 && sendPacket == 0) {
+                    // Write samples to audio buffer
+
+                    for (size_t i = 0;
```

Comment:
Is there any way to extract this inner loop so it can be shared between all of the frame handling paths?

Reply:
Discussed on IRC: task to do in later pull.